### PR TITLE
allow salt to be nullable - compatibility with bcrypt encoder

### PR DESCRIPTION
**Pull request by JarekW**

### CODE REVIEW

File: Resources/doc/installation.md [hunk 1]
```diff
@@ -150,4 +150,20 @@ security:
         - { path: ^/admin, roles: ROLE_ADMIN }
 ```
 
+### Bcrypt encoder
+
+If you want to use bcrypt encoder, you must clear salt that is set in user constructor, like this:
```
Comment (szymach) [inline Resources/doc/installation.md]: *the salt

**Comment by JarekW** [comment]
@fsi-open

**Comment by szymach** [comment]
:+1:

**Comment by JarekW** [comment]
@rn0 :)

**Comment by rn0** [comment]
:+1:

File: Resources/doc/installation.md [hunk 2]
```diff
@@ -150,4 +150,20 @@ security:
         - { path: ^/admin, roles: ROLE_ADMIN }
 ```
 
+### Bcrypt encoder
+
+If you want to use bcrypt encoder, you must clear the salt that is set in user constructor, like this:
```
Comment (chives) [inline Resources/doc/installation.md]: We should mention that this notice is important for SF >= 2.8

**Comment by norzechowicz** [comment]
why?

**Comment by JarekW** [comment]
[2.7](https://github.com/symfony/security/blob/2.7/Core/Encoder/BCryptPasswordEncoder.php#L81)
[2.8](https://github.com/symfony/security/blob/2.8/Core/Encoder/BCryptPasswordEncoder.php#L79)
[3.0](https://github.com/symfony/security/blob/3.0/Core/Encoder/BCryptPasswordEncoder.php#L77)

in 2.8 salt passed to `BCryptPasswordEncoder` starts to be deprecated, in 3.0 salt is ignored. This can lead to case when we have salt in column, but it is not used/connected with password in another column (in some algorithms like bcrypt). I found that when i worked wih behat - it somehow cathed `trigger_error` from code in 2.8.

**Comment by chives** [comment]
:+1:

**Comment by norzechowicz** [comment]
this might also lead to a potential security issue when somebody who is not using BCrypt forgot about salt, right?

**Comment by chives** [comment]
That's what the clarification in the docs is for. It would be security issue if someone overwrites the salt to null and use encoder other than bcrypt.

**Pull request merged**